robot/packages/deferred_package_manager.go

```diff
@@ -0,0 +1,175 @@
+// instantly and behaves as a noop package manager until a connection to app has been established.
+//
+// Raison d'être: AquireConnection will waste 5 seconds timing out on robots that have no internet connection but have downloaded
+// all of their cloud_packages
+//
+// This puts an optimization on top of that behavior: On the first start of the package manager, if all of the expected packages are
+// present on the system, put the app-connection establishment in a goroutine and use a noopManager for the first Sync.
+// If there are missing packages, this will still block to prevent a half-started robot.
+
+import (
+	"context"
+	"os"
+	"sync"
+
+	"github.com/pkg/errors"
+	pb "go.viam.com/api/app/packages/v1"
+	goutils "go.viam.com/utils"
+
+	"go.viam.com/rdk/config"
+	"go.viam.com/rdk/logging"
+	"go.viam.com/rdk/resource"
+)
+
+type deferredPackageManager struct {
+	resource.Named
+	resource.TriviallyReconfigurable
+	ctx                 context.Context
+	establishConnection func(ctx context.Context) (pb.PackageServiceClient, error)
+	cloudManager        ManagerSyncer
+	cloudManagerArgs    cloudManagerConstructorArgs
+	cloudManagerLock    sync.Mutex
+
+	lastSyncedManager     ManagerSyncer
+	lastSyncedManagerLock sync.Mutex
+
+	logger logging.Logger
+}
+type cloudManagerConstructorArgs struct {
+	cloudConfig *config.Cloud
+	packagesDir string
+	logger      logging.Logger
+}
+
+var (
+	_ Manager       = (*deferredPackageManager)(nil)
+	_ ManagerSyncer = (*deferredPackageManager)(nil)
+)
+
+// DeferredServiceName is used to refer to/depend on this service internally.
+var DeferredServiceName = resource.NewName(resource.APINamespaceRDKInternal.WithServiceType(SubtypeName), "deferred-manager")
+
+// NewDeferredPackageManager returns a DeferredPackageManager. See deferred_package_manager.go for more details.
+func NewDeferredPackageManager(
+	ctx context.Context,
+	establishConnection func(ctx context.Context) (pb.PackageServiceClient, error),
+	cloudConfig *config.Cloud,
+	packagesDir string,
+	logger logging.Logger,
+) ManagerSyncer {
+	noopManager := noopManager{Named: InternalServiceName.AsNamed()}
+	return &deferredPackageManager{
+		Named:               DeferredServiceName.AsNamed(),
+		ctx:                 ctx,
+		establishConnection: establishConnection,
+		cloudManagerArgs:    cloudManagerConstructorArgs{cloudConfig, packagesDir, logger},
+		lastSyncedManager:   &noopManager,
+		logger:              logger,
+	}
+}
+
+// Sync syncs packages and removes any not in the list from the local file system.
+// If there are packages missing on the local fs, this will wait while attempting to establish a connection to app.viam.
+//
+// Sync is the core state-setting operation of the package manager so if we sync with one manager,
+// all subsequent operations should use the same manager until the next sync.
+func (m *deferredPackageManager) Sync(ctx context.Context, packages []config.PackageConfig) error {
```

Comment:
Nice I think this solves the offline<->online transitions well too:

Offline + Packages correct: A noop manager is created. When robot gets new config and packages change, LocalRobot.Reconfigure() happens and Sync is recalled, triggering a switch from noop to cloud.
Online + Packages correct: Same as a Offline + Packages correct
Offline + Packages incorrect: A cloud manager is attempted to created immediately, then A thread is created to retry connection infinitely until a connection can be established, then packages are synced. The robot will start whatever modules it can and fail on the ones where files aren't present.
Online + Packages are incorrect: A cloud manager is attempted to created immediately.

Reply:
I know that I said "assume package download has completed successfully" when discussing yesterday. But one case to think through:

1. Robot is online. Packages are being synced due to reconfigure.
2. Robot goes offline in the middle of package download.
3. Package manager is still in cloud mode, and downloads fail.
4. Robot comes back online, but a reconfigure doesn't happen again because no config changes.

I think in this case the robot is stuck in a bad state and has to be restarted? I think we might want to have a routine higher up that detects and transition from offline<->online and calls Sync to reevaluate?

Reply:
I think this is what abe was talking about in https://github.com/viamrobotics/rdk/pull/3677#discussion_r1522309762

I played around for ~20 minutes and tried to replicate this error case (forcing the package manager to try to download packages every reconfigure) but the gRPC connection seemed to come online again ?? Someone online claimed that go's gRPC library handles reconnects but I searched through the code and wasn't able to find solid evidence of that. It was very late at night and was tired so take all of this with a grain of salt.

However, even if this error case does exist, the solution would complicate this logic a fair bit and I'd rather do it in a future PR (if everyone involved is okay with that) (this is not the problem I set out to solve, but I did lay the groundwork to help solve it)

Reply:
Ok sounds good, future PR and we will track it in https://viam.atlassian.net/browse/RSDK-6880